### PR TITLE
Iterate in reverse when comparing Stacks

### PR DESCRIPTION
**Pull request by saethlin**
I was originally going to propose this after I (eventually, sorry) am done with https://github.com/rust-lang/miri/pull/1935 but @nilstrieb independently identified this code path as hot and came up with a different optimization, which is basically this: https://github.com/rust-lang/rust/issues/49892#issuecomment-1140613283

Anecdotally, I've found this optimization to be much more effective but of course that is probably workload-dependent.

### CODE REVIEW

File: src/stacked_borrows.rs [hunk 1]
```diff
@@ -88,6 +88,17 @@ pub struct Stack {
     borrows: Vec<Item>,
 }
 
+/// This implementation is primarily used when attempting to merge borrow stacks for adjacent
+/// bytes. For adjacent stacks, when the stacks differ at all, they tend to differ either in
+/// length or at the end of the borrows array. Iterating in reverse returns faster for `Stack`s
+/// that are not equal.
+impl PartialEq for Stack {
+    fn eq(&self, other: &Self) -> bool {
+        self.borrows.len() == other.borrows.len()
```
Comment (RalfJung) [inline src/stacked_borrows.rs:97]: Can you do this via
```
let Stack { borrows } = self;
```
to ensure this fails to compile when we add more fields?

**Comment by RalfJung** [comment]
Thanks, this makes sense!

Bummer that we don't have perf tracking infrastructure... I'll just trust that this works.

**Comment by saethlin** [comment]
We could also wait for @Nilstrieb to confirm that this is at least as impactful as they optimization they were considering.

**Comment by RalfJung** [comment]
Well, have you tested this PR and confirmed that it helps on some benchmarks?

**Comment by Nilstrieb** [comment]
Hmm, that is interesting. Running the `serde1` benchmark with `hyperfine` results in this:
```
faster-tag-partial-eq (local)
Benchmark 1: cargo +miri miri run
  Time (mean ± σ):      5.954 s ±  0.360 s    [User: 5.862 s, System: 0.087 s]
  Range (min … max):    5.670 s …  6.845 s    10 runs

master
Benchmark 1: cargo +miri miri run
  Time (mean ± σ):      6.160 s ±  0.069 s    [User: 6.078 s, System: 0.076 s]
  Range (min … max):    6.080 s …  6.313 s    10 runs

faster-stack-cmp
Benchmark 1: cargo +miri miri run
  Time (mean ± σ):      6.202 s ±  0.087 s    [User: 6.114 s, System: 0.080 s]
  Range (min … max):    6.042 s …  6.332 s    10 runs
```

**Comment by Nilstrieb** [comment]
Rebasing the two and comparing them with master again
```
both serde1
Benchmark 1: cargo +miri miri run
  Time (mean ± σ):      5.260 s ±  0.106 s    [User: 5.158 s, System: 0.079 s]
  Range (min … max):    5.201 s …  5.554 s    10 runs

both mse
Benchmark 1: cargo +miri miri run
  Time (mean ± σ):     30.479 s ±  0.226 s    [User: 30.155 s, System: 0.317 s]
  Range (min … max):   30.340 s … 31.095 s    10 runs

master serde1
Benchmark 1: cargo +miri miri run
  Time (mean ± σ):      6.152 s ±  0.062 s    [User: 6.071 s, System: 0.074 s]
  Range (min … max):    6.069 s …  6.268 s    10 runs

master mse
Benchmark 1: cargo +miri miri run
  Time (mean ± σ):     35.022 s ±  0.408 s    [User: 34.702 s, System: 0.313 s]
  Range (min … max):   34.658 s … 36.133 s    10 runs
```

**Comment by RalfJung** [comment]
Those are nice wins, but what I meant is comparing this PR with master after https://github.com/rust-lang/miri/pull/2218 landed. The wins you are showing might be entirely due to https://github.com/rust-lang/miri/pull/2218?

**Comment by saethlin** [comment]
Based on some experimentation I just did on my end, I think the thing to do here is land a few of my local test cases as `bench-cargo-miri` programs as well as land https://github.com/rust-lang/miri/pull/1935 then re-evaluate micro-optimizations in this area.

**Comment by saethlin** [comment]
I'll re-open this if it still seems interesting after making all the decisions on https://github.com/rust-lang/miri/pull/2315

**Comment by saethlin** [comment]
Since https://github.com/rust-lang/miri/pull/2315, the stack comparison code path is now optimized to a `memcmp`, and in combination with fact that we are comparing less bits makes this code path a lot less interesting. Awesome!

**Pull request closed without merging**